### PR TITLE
fix(modal): handle removal of open attribute and prevent multiple beforeClose calls

### PR DESCRIPTION
**Pull request by driskull**
**Related Issue:** #6407 #6379

## Summary

- Add internal property `opened` to maintain open state when beforeBack is rejected and to handle initial animation.
- Refactor `beforeClose` default value.
- Rename internal close method name to be consistent with internal open method name
- Prevent `beforeClose` from being called twice by adding an internal flag to ignore watch changes when necessary.
- Tests

### CODE REVIEW

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 1]
```diff
@@ -123,14 +123,40 @@ describe("calcite-modal properties", () => {
         (elm.beforeClose = (window as typeof window & Pick<typeof elm, "beforeClose">).beforeClose)
     );
     await page.waitForChanges();
-    await modal.setProperty("open", true);
```
Comment (alisonailea) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:126]: is there any way to fix this without changing the property name? A prop name change is technically a breaking change.

Comment (driskull) [reply packages/calcite-components/src/components/modal/modal.e2e.ts:126]: This isn't a breaking change, no prop is being removed.

**Review by alisonailea — APPROVED** [review]

**Comment by driskull** [comment]
@jcfranco can you review this one?

File: packages/calcite-components/src/components/modal/modal.tsx [hunk 2]
```diff
@@ -343,6 +342,15 @@ export class Modal
   //
   //--------------------------------------------------------------------------
 
+  /**
+   * We use an internal property to handle styles for when a modal is actually opened, not just when the open attribute is applied. This is a property because we need to apply styles to the host element and to keep the styles present while beforeClose is .
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.tsx:346]: Nitpicks: 

* `...while beforeClose is .` ➡️  `...while beforeClose is.`
* `@internal.` ➡️ `@internal`

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 3]
```diff
@@ -123,14 +123,40 @@ describe("calcite-modal properties", () => {
...
+  const page = await newE2EPage();
+  const mockCallBack = jest.fn();
+  await page.exposeFunction("beforeClose", mockCallBack);
+  await page.setContent(`
+    <calcite-modal open></calcite-modal>
+  `);
+  const modal = await page.find("calcite-modal");
+  await page.$eval(
+    "calcite-modal",
+    (elm: HTMLCalciteModalElement) =>
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:147]: Nitpick: `elm` ➡️ `el` for consistency.

Sidebar, I think we can have ESLint catch/fix this one.

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 4]
```diff
@@ -123,14 +123,40 @@ describe("calcite-modal properties", () => {
...
+      (elm.beforeClose = (window as typeof window & Pick<typeof elm, "beforeClose">).beforeClose)
+  );
+  await page.waitForChanges();
+  modal.setProperty("open", true);
+  await page.waitForChanges();
+  expect(await modal.getProperty("opened")).toBe(true);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:153]: `opened` is internal, so maybe we can avoid asserting on it. I'll leave this up to you.

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 5]
```diff
@@ -123,14 +123,40 @@ describe("calcite-modal properties", () => {
...
+  const mockCallBack = jest.fn();
+  await page.exposeFunction("beforeClose", mockCallBack);
+  await page.setContent(`
+    <calcite-modal open></calcite-modal>
+  `);
+  const modal = await page.find("calcite-modal");
+  await page.$eval(
+    "calcite-modal",
+    (elm: HTMLCalciteModalElement) =>
+      (elm.beforeClose = (window as typeof window & Pick<typeof elm, "beforeClose">).beforeClose)
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:148]: Suggestion: I _think_ you could use our global test type for this `GlobalTestProps<{ beforeClose: HTMLCalciteModalElement["beforeClose"] }>`.

File: packages/calcite-components/src/components/modal/modal.tsx [hunk 6]
```diff
@@ -343,6 +342,15 @@ export class Modal
...
+   *
+   * @internal.
+   */
+  @Prop({ mutable: true, reflect: true }) opened = false;
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.tsx:350]: For consistency, can we move this to properties despite it being internal?

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 7]
```diff
@@ -109,7 +109,7 @@ describe("calcite-modal properties", () => {
     expect(styleH).toEqual("800px");
   });
 
-  it("calls the beforeClose method prior to closing", async () => {
+  it("calls the beforeClose method prior to closing via click", async () => {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:112]: Can you add a similar test covering closing via `Escape` key?

File: packages/calcite-components/src/components/modal/modal.tsx [hunk 8]
```diff
@@ -87,8 +87,7 @@ export class Modal
   @Prop({ mutable: true, reflect: true }) open = false;
 
   /** Passes a function to run before the component closes. */
-  @Prop()
-  beforeClose: (el: HTMLElement) => Promise<void> = () => Promise.resolve();
+  @Prop() beforeClose: (el: HTMLCalciteModalElement) => Promise<void>;
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.tsx:90]: For posterity, this is removing the default value, but we should be OK as it's [documented as `undefined`](https://developers.arcgis.com/calcite-design-system/components/modal/#api-reference-properties-beforeClose).

**Review by jcfranco — APPROVED** [review]
Awesome! Nice solution and clean to boot! ✨🥾✨

File: packages/calcite-components/src/components/modal/modal.e2e.ts (hunk 3 above)
Comment (driskull) [reply packages/calcite-components/src/components/modal/modal.e2e.ts:147]: Replaced all var names. I think someone else named these but yeah eslint would be good.

File: packages/calcite-components/src/components/modal/modal.e2e.ts (hunk 4 above)
Comment (driskull) [reply packages/calcite-components/src/components/modal/modal.e2e.ts:183]: I'll leave it as it ensures the prop is being added.

**Pull request merged**